src/components/date/date.jsx

```diff
@@ -0,0 +1,264 @@
+    setExpanded(!expanded);
+  }
+
+  function onCalendarChange(newDate) {
+    setValue(newDate);
+    setEditing(false);
+    setExpanded(false);
+  }
+
+  useDropdownClose(containerRef, expanded, () => setExpanded(false));
+
+  useEffect(() => {
+    setEditing(!!props.validationMessage);
+    setValidationMessage(props.validationMessage);
+  }, [props.validationMessage]);
+
+  useEffect(() => {
+    setValue(fromFullDateFormat(props.value));
+  }, [props.value]);
+
+  useEffect(() => {
+    if (active && props.onChange) props.onChange({ target: ref.current });
+  }, [value]);
+
+  useLayoutEffect(() => {
+    if (active) inputRef.current.focus();
+  }, [editing]);
+
+  useLayoutEffect(() => {
+    inputRef.current.setCustomValidity(validationMessage);
+  }, [validationMessage]);
+
+  useLayoutEffect(() => {
+    inputRef.current.value = editing ? toFullDateFormat(value) : toDateStringFormat(value);
+  }, [value]);
+
+  useImperativeHandle(ref, () => ({
+    get dirty() {
+      const providedValue = props.value;
+      return providedValue !== this.value;
+    },
+    id: props.id,
+    name: props.name,
+    get value() {
+      return toFullDateFormat(value);
+    },
+  }));
+
+  return (
+    <div
+      className={classNames.layout.container}
+      onBlur={onBlur}
+      ref={containerRef}
+    >
+      <FormControl
+        error={validationMessage}
+        id={ids.input}
+        labelId={ids.label}
+        label={props.label}
+        readOnly={props.readOnly}
+        required={props.required}
+      >
+        <input
+          aria-describedby={ids.validationMessage}
+          className={classNames.input}
+          defaultValue={editing ? toFullDateFormat(value) : toDateStringFormat(value)}
+          id={ids.input}
+          key={`${ids.input}-${editing ? 'editing' : 'display'}`}
+          max={props.max}
+          min={props.min}
+          name={props.name}
+          onChange={onInputChange}
+          onClick={onInputClick}
+          onKeyDown={onInputKeyDown}
+          placeholder={props.placeholder}
+          readOnly={props.readOnly}
+          ref={inputRef}
+          required={props.required}
+          type={editing ? 'date' : 'text'}
+        />
+        <ControlIcons
+          label={props.label}
+          onPress={onIconPress}
+          readOnly={props.readOnly}
+          validationMessage={validationMessage}
+        />
+      </FormControl>
+      {expanded && (
+        <div className={classNames.layout.calendar}>
+          <Calendar onDateSelected={onCalendarChange} value={toFullDateFormat(value)} />
+        </div>
+      )}
+    </div>
+  );
+});
+
+Date.propTypes = {
+  classNames: PropTypes.shape({
+    input: PropTypes.string,
+    layout: PropTypes.shape({
+      container: PropTypes.string,
+      calendar: PropTypes.string,
+    }),
+  }),
+  id: PropTypes.string.isRequired,
+  label: PropTypes.string.isRequired,
+  /** The latest date to accept in full-date format (i.e. yyyy-mm-dd) */
```

Comment:
I think it would be interesting to explore our own validation function for these date props!

Reply:
Good idea! I'll write a story to create a prop-types folder and make the first one called `PropTypes.fullDate`

Reply:
https://www.pivotaltracker.com/story/show/177847305